### PR TITLE
Enumerating acquisition paths

### PR DESCRIPTION
**Pull request by nuclearkatie**
`acquisition_paths` builds a DiGraph of the fuel cycle and then enumerates paths through it

`pathway_analysis` includes a variety of tools that can be called in the command line version of trailmap to analyze and simply the acquisition pathways

### CODE REVIEW

File: trailmap/pathway_analysis.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+def draw_path(G, path):
+    draw_graph(G)
+    pos = nx.drawing.nx_pydot.graphviz_layout(G)
+    nx.draw_networkx_nodes(G, pos=pos, nodelist=path, node_color="red")
+    plt.show()
+    return
+
+
+def print_graph_parameters(G):
```
Comment (gonuke) [inline trailmap/pathway_analysis.py:28]: might be interesting to brainstorm on parameters of interest:
* depth of graph (i.e. longest path)
* shortest path
* number of pathways

File: trailmap/pathway_analysis.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    max_flow = maximum_flow_value(G, s, t)
+    return max_flow_path, max_flow
+
+
+def find_simple_cycles(G):
+    sc = list(nx.simple_cycles(G))
+    return sc
+
+
+def find_facility_specific_paths(pathways, facility, type='any'):
```
Comment (gonuke) [inline trailmap/pathway_analysis.py:66]: `type` already means something in python, so you may want a different variable name to avoid confusion/conflict

File: trailmap/pathway_analysis.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+
+def find_paths_containing_facilities(pathways, facilities, type='any'):
+    p = []
+    if type == 'any':
+        for path in pathways:
+            p.append([path for facility in facilities if facility in path])
+    if type == 'all':
+        for path in pathways:
+            if set(facilities).issubset(path) is True:
```
Comment (gonuke) [inline trailmap/pathway_analysis.py:99]: Don't usually need to test `is True`.  i.e.  `3 == 3 is True` is the same as `3 == 3`
```suggestion
            if set(facilities).issubset(path):
```

File: trailmap/pathway_analysis.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    return p
+
+
+def find_shortest_path(pathways):
+    '''finds the pathway with the shortest number of steps from source to
+    target. Returns a tuple with path and length.
+    '''
+    if len(pathways) is not 0:
+        shortest_path_length = min([len(path) for path in pathways])
+        shortest_path = min([path for path in pathways])
```
Comment (gonuke) [inline trailmap/pathway_analysis.py:111]: Have you tested this?  I'm not sure it works the way you think it does...

Also, while this is a compact notation, it requires you to loop over pathways twice instead of once.

File: trailmap/acquisition_paths.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+def build_graph(facility_dict_in, facility_dict_out):
+    '''Builds NetworkX graph from Cyclus input file
+    '''
+    G = nx.DiGraph()
+
+    for facility in facility_dict_in.keys():
+        G.add_node(facility)
+    for facility in facility_dict_in:
+        for incommod in facility_dict_in[facility]:
```
[the remark this answers is not in the document]
Comment (nuclearkatie) [reply trailmap/acquisition_paths.py:28]: I like that suggestion, thanks

File: trailmap/pathway_analysis.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    pos = nx.drawing.nx_pydot.graphviz_layout(G, prog=dot)
+
+    plt.figure()
+    # nx.draw_shell(G, with_labels=True)
+    nx.draw_networkx_nodes(G, pos)
+    nx.draw_networkx_labels(G, pos)
+    nx.draw_networkx_edges(G, pos)
+    # nx.draw_networkx_edge_labels(G, pos)
+
+    return plt
```
[the remark this answers is not in the document]
Comment (nuclearkatie) [reply trailmap/pathway_analysis.py:17]: No idea why it's like that. Probably some holdover from throwing things at the wall, will remove

File: trailmap/acquisition_paths.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (nuclearkatie) [reply trailmap/acquisition_paths.py:28]: The loops are different because they're doing different things. For the receiver, you're iterating through every incommod in the list of incommods to be able to find senders that can provide that commod. But when you're looping through senders, you just need to check that the commod is in the list oucommods. You don't need to iterate through outcommods, because you don't care about the other outcommods So even if I changed the loop, I would still need to iterate over incommods directly later to get individual commods. That's why I wrote it the way I did

I'm sure there's a better way of doing this, but I'm not sure your suggestion is better per sae because it won't reduce the number of nested loops

File: trailmap/pathway_analysis.py (hunk 1 above)
Comment (nuclearkatie) [reply trailmap/pathway_analysis.py:28]: Yeah those are so simple that it doesn't make sense to make people specifically request them, you're right I should just print those as part of the general graph parameters. Even that I currently have set as a command line flag, do you think it's important enough that I should just remove that flag and auto-print it?

File: trailmap/acquisition_paths.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+
+def find_simple_paths(G, sources, sinks):
+    ''' finds all simple paths between a given list of sources and targets
+    '''
+    pathways = set()
+    for source, sink in [(x, y) for x in sources for y in sinks]:
+        if source in G and sink in G:
+            paths = list(nx.all_simple_paths(G, source=source, target=sink))
+            [pathways.add(tuple(path)) for path in paths]
```
[the remark this answers is not in the document]
Comment (nuclearkatie) [reply trailmap/acquisition_paths.py:55]: Do you have an idea for how to do that? I think you can only run `all_simple_paths` between a single source and target, so I figured looping over each source and target and then adding the resulting pathways for those specific facilities to the overall list would be a good way to go about it, but I can see how that would be confusing.

[the remark this answers is not in the document]
Comment (gonuke) [reply trailmap/acquisition_paths.py]: I think I just meant something like this:
```
for path in paths:
   pathways.add(tuple(path))
```

[the remark this answers is not in the document]
Comment (nuclearkatie) [reply trailmap/acquisition_paths.py]: Ah I thought you were proposing I try to get rid of `paths` entirely and was a bit confused!

File: trailmap/pathway_analysis.py (hunk 2 above)
Comment (nuclearkatie) [reply trailmap/pathway_analysis.py:74]: Good catch. I was calling it `location` in `main` already anyway

File: trailmap/pathway_analysis.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+                subset_pathways.append(path)
+            else:
+                subset_pathways.append(path)
+
+    return subset_pathways
+
+
+def find_paths_between_facilities(pathways, source, target):
+    subset_pathways = []
+    for path in (x for x in pathways if x[0] == source and x[-1] == target):
```
[the remark this answers is not in the document]
Comment (nuclearkatie) [reply trailmap/pathway_analysis.py:94]: This function was written specifically to find pathways with a given source and target node, so yes (A, C) wouldn't return anything if C was not the end of any path. However the next function, `find_paths_containing_facilities`, would find any path with A and C regardless of where along the path they're located

File: trailmap/acquisition_paths.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    return G, pathways
+
+
+def build_graph(facility_dict_in, facility_dict_out):
+    '''Builds NetworkX graph from Cyclus input file
+    '''
+    G = nx.DiGraph()
+    G.add_nodes_from(facility_dict_in.keys())
+    for reciever in facility_dict_in:
+        for incommod in facility_dict_in[reciever]:
```
Comment (gonuke) [inline trailmap/acquisition_paths.py:26]:
```suggestion
    for reciever, incommods in facility_dict_in.items():
        for incommod in incommods:
```

File: trailmap/acquisition_paths.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    return acquisition_paths
+
+
+def find_simple_paths(G, sources, sinks):
+    ''' finds all simple paths between a given list of sources and targets
+    '''
+    pathways = set()
+    for source, sink in [(x, y) for x in sources for y in sinks]:
+        paths = list(nx.all_simple_paths(G, source=source, target=sink))
+        [pathways.add(tuple(path)) for path in paths]
```
Comment (gonuke) [inline trailmap/acquisition_paths.py:52]: (It seems the comment got lost somewhere, but...)
Maybe this is a more pythonic way to do this?
```suggestion
        pathways.update([tuple(path) for path in paths])
```

File: trailmap/pathway_analysis.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (gonuke) [reply trailmap/pathway_analysis.py:94]: Ultimately might want another function to take the results of `find_paths_containing_facilities` and filter for those which have A before C in the path

File: trailmap/acquisition_paths.py (hunk 9 above)
Comment (nuclearkatie) [reply trailmap/acquisition_paths.py:26]: This does not work as intended, trying to figure out why

File: trailmap/acquisition_paths.py (hunk 10 above)
Comment (nuclearkatie) [reply trailmap/acquisition_paths.py:52]: I was going to do 
`for path in paths:
            pathways.add(tuple(path))`

File: trailmap/pathway_analysis.py (hunk 2 above)
Comment (nuclearkatie) [reply trailmap/pathway_analysis.py]: fixed in more recent commit

File: trailmap/pathway_analysis.py (hunk 4 above)
Comment (gonuke) [reply trailmap/pathway_analysis.py:119]:
```suggestion
        shortest_path = [path for path in pathways if len(path) == shortest_length]
```

**Comment by bam241** [comment]
you have a typo: `reciever` -> `receiver`

File: trailmap/acquisition_paths.py (hunk 9 above)
Comment (nuclearkatie) [reply trailmap/acquisition_paths.py]: Figured it out, incorporated your comment. Thanks!

File: trailmap/pathway_analysis.py (hunk 4 above)
Comment (nuclearkatie) [reply trailmap/pathway_analysis.py]: pathway analysis has been taken out of this PR, but I did make that change. Thanks!

**Comment by nuclearkatie** [comment]
Whew! Tests have been written, all comments have either been incorporated, addressed, or pushed until the next PR (which will include the pathway analysis stuff that was originally in this PR). This PR now covers only the creation of a graph and the generation of acquisition paths.

Hopefully now it is ready for one more check and then a merge...? @gonuke and maybe @bam241 if you'd like to look it over as well

**Comment by nuclearkatie** [comment]
Closes #27

File: scripts/main.py [hunk 11]
```diff
@@ -1,13 +1,24 @@
 import argparse
 import trailmap.parse_input as pi
 import trailmap.commodity_dictionary as cd
```
Comment (bam241) [inline scripts/main.py:3]: local import should be located after global ones

File: scripts/main.py [hunk 12]
```diff
@@ -19,8 +30,19 @@ def main(args=None):
...
                                              commodity_dictionary)
+        (G, pathways) = ap.conduct_apa(facility_dict_in,
+                                       facility_dict_out)
+        # pa.print_graph_parameters(G, pathways)
+
+        # if ns.draw:
+        #     plt = pa.draw_graph(G)
+        #     plt.show()
+        # if ns.pickle:
+        #     nx.write_gpickle(G, ns.picklefile)
```
Comment (bam241) [inline scripts/main.py:44]: do you need to keep those comments ?

File: scripts/main.py [hunk 13]
```diff
@@ -1,13 +1,24 @@
 import argparse
 import trailmap.parse_input as pi
 import trailmap.commodity_dictionary as cd
+import trailmap.acquisition_paths as ap
+# import trailmap.pathway_analysis as pa
+import networkx as nx
+from pprint import pprint
+import matplotlib.pyplot as plt
```
Comment (bam241) [inline scripts/main.py:8]: is matplotlib actually required ? and sprint ?

File: scripts/main.py [hunk 14]
```diff
@@ -1,13 +1,24 @@
...
 
 
 def make_parser():
     """Makes the Cyclus-Trailmap command line parser"""
     p = argparse.ArgumentParser(description="Cyclus-Trailmap command line",
                                 epilog="python main.py cyclus_input_file.xml")
     p.add_argument('infile', nargs=1, help='Cyclus input file. Must be XML')
```
Comment (bam241) [inline scripts/main.py:15]: you can add a required option to the parameter:
`required - Whether or not the command-line option may be omitted (optionals only).`  [[see](https://docs.python.org/3/library/argparse.html)]

File: scripts/main.py [hunk 15]
```diff
@@ -19,8 +30,19 @@ def main(args=None):
 
     if ns.infile[0] is not None:
```
Comment (bam241) [inline scripts/main.py:31]: if statement should not be required if you add the `required` option in the argparse...

File: tests/test_acquisition_paths.py [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+
+@pytest.mark.parametrize("name, edges, fd_in, fd_out, exp_paths", testdata)
+def test_build_graph(name, edges, fd_in, fd_out, exp_paths):
+    exp_G = nx.MultiDiGraph()
+    exp_G.add_edges_from(edges)
+    obs_G = ap.build_graph(fd_in, fd_out)
+
+    edit_dist = nx.graph_edit_distance(exp_G, obs_G)
+    assert edit_dist == 0.0
```
Comment (bam241) [inline tests/test_acquisition_paths.py:17]:
```suggestion
    assert edit_dist == nx.graph_edit_distance(exp_G, obs_G)
```

File: trailmap/acquisition_paths.py [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+def conduct_apa(facility_dict_in, facility_dict_out):
+    ''' '''
+    G = build_graph(facility_dict_in, facility_dict_out)
+
+    sources = list(node for node, in_deg in G.in_degree() if in_deg == 0)
+    sinks = list(node for node, out_deg in G.out_degree() if out_deg == 0)
+
+    print("\nSimple paths")
+    pathways = find_simple_paths(G, sources, sinks)
+    print(pathways)
```
Comment (bam241) [inline trailmap/acquisition_paths.py:14]: are those print statement required ? (or what is just for debugging ?)

File: trailmap/acquisition_paths.py [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+
+
+def find_simple_paths(G, sources, sinks):
+    ''' finds all simple paths between a given list of sources and targets
+    '''
+    pathways = set()
+    for source, sink in [(x, y) for x in sources for y in sinks]:
+        paths = list(nx.all_simple_paths(G, source=source, target=sink))
+        for path in paths:
+            pathways.add(tuple(path))
```
Comment (bam241) [inline trailmap/acquisition_paths.py:40]: maybe you can inline this one ?

File: trailmap/acquisition_paths.py [hunk 19]
```diff
@@ -0,0 +1,42 @@
+import networkx as nx
+from pprint import pprint
```
Comment (bam241) [inline trailmap/acquisition_paths.py:2]: `pprint` seemed not to be used

**Review by bam241 — COMMENTED** [review]
Looking good !

some few syntax comments, nothing major.

File: tests/test_acquisition_paths.py (hunk 16 above)
Comment (gonuke) [reply tests/test_acquisition_paths.py:17]: I think you mean
```
assert nx.graph_edit_distance(exp_G, obs_G) == 0
```

File: trailmap/acquisition_paths.py [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+
+def build_graph(facility_dict_in, facility_dict_out):
+    '''Builds NetworkX graph from Cyclus input file
+    '''
+    G = nx.MultiDiGraph()
+    G.add_nodes_from(facility_dict_in.keys())
+    for receiver, incommods in facility_dict_in.items():
+        # find all facilities with that commod as an outcommod
+        for sender, outcommods in facility_dict_out.items():
+            connection = list(set(incommods).intersection(outcommods))
```
Comment (gonuke) [inline trailmap/acquisition_paths.py:27]: This probably doesn't need to be a `list` - it should work as a `set`

(not super important, but a little cleaner)

File: trailmap/acquisition_paths.py (hunk 17 above)
Comment (gonuke) [reply trailmap/acquisition_paths.py:14]: If the print statements are intended for something other than debugging... then we should discuss separation of calculation from I/O.  Ideally, this method should just return the pathways and let another method take care of output

File: trailmap/acquisition_paths.py [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+                G.add_edge(sender, receiver, commodity=commod)
+    return G
+
+
+def find_simple_paths(G, sources, sinks):
+    ''' finds all simple paths between a given list of sources and targets
+    '''
+    pathways = set()
+    for source, sink in [(x, y) for x in sources for y in sinks]:
+        paths = list(nx.all_simple_paths(G, source=source, target=sink))
```
Comment (gonuke) [inline trailmap/acquisition_paths.py:38]: Could this just be:
```suggestion
        pathways = set(nx.all_simple_paths(G, source=source, target=sink))
```
or
```suggestion
        pathways = set([tuple(path) for path in nx.all_simple_paths(G, source=source, target=sink)])
```

Is the tuple-ness important (I seem to remember that it is but can't remember why).

File: trailmap/acquisition_paths.py (hunk 20 above)
Comment (gonuke) [reply trailmap/acquisition_paths.py:27]: Could also skip assigning the variable and just have
```
for commod in set(incommods).intersection(outcommods):
```

**Review by gonuke — CHANGES_REQUESTED** [review]
This looks pretty solid.  I have a few style comments, some following up on @bam241's

File: trailmap/acquisition_paths.py (hunk 18 above)
Comment (nuclearkatie) [reply trailmap/acquisition_paths.py:40]: I actually had this as a single line in an older/first version
`[pathways.update(tuple(path)) for path in paths]`
@gonuke recommended I break it into multi-line for clarity

File: scripts/main.py (hunk 12 above)
Comment (nuclearkatie) [reply scripts/main.py:44]: Sloppy on my part, its a holdover from when this PR included pathway analysis stuff. I guess for cleanliness I should remove the text entirely and then bring it back in a future PR where that code will actually do something

File: scripts/main.py (hunk 13 above)
Comment (nuclearkatie) [reply scripts/main.py:8]: Another dependency that actually relates to the code I pulled out or this PR. Nothing currently in this PR uses it

File: trailmap/acquisition_paths.py (hunk 18 above)
Comment (bam241) [reply trailmap/acquisition_paths.py:40]: I just found it silly to loop over the `paths` for append each element to the pathways.

Maybe you could convert paths into a tuple and append that topple directly ?

Comment (bam241) [inline trailmap/acquisition_paths.py:40]:
```suggestion
    for source, sink in [(x, y) for x in sources for y in sinks]:
        pathways += tupple(nx.all_simple_paths(G, source=source, target=sink))
```

what about ?

Comment (nuclearkatie) [reply trailmap/acquisition_paths.py:40]: I don't want a tuple of paths in pathways though.
`list(nx.all_simple_paths(G, source=source, target=sink))` returns a list of lists, for example 
`[['SourceA', 'SinkA'], ['SourceA', 'Facility', 'SinkA']]. I want to turn each individual element in that list into a tuple and add it to pathways set

Let's say `pathways = {('SourceB', 'SinkB')}` and now we're iterating over source=SourceA and sink=SinkA. The output of `list(nx.all_simple_paths(G, source=source, target=sink))` is `[['SourceA', 'SinkA'], ['SourceA', 'Facility', 'SinkA']]`, a list of lists. Turning that into a tuple would give `(['SourceA', 'SinkA'], ['SourceA', 'Facility', 'SinkA'])`, which isn't what I want, and can't be added anyway because it still includes an unhashable type. I want each element to become a tuple so it can be added to the set of tuples `pathways`.

In this mini-example, what I want to reach is {('SourceB', 'SinkB'), ('SourceA', 'SinkA'), ('SourceA', 'Facility', 'SinkA')}.

I don't really have a preference between `[pathways.update(tuple(path)) for path in paths]` and 
```for path in paths:
            pathways.add(tuple(path))```
but I also don't see a clear other way to add elements from a list of lists as individual tuples to a set

File: trailmap/acquisition_paths.py (hunk 21 above)
Comment (nuclearkatie) [reply trailmap/acquisition_paths.py:38]: you can't take the output of `nx.all_simple_paths(G, source=source, target=sink)` and put it into a set because the output is a list of lists, and you get an unhashable error because you're effectively trying to turn a list of lists into a set of lists (not allowed)

tuples for individual pathways are important because they allow elements to be used again within the tuple but don't allow duplicates of the entire tuple. Now I guess technically you don't need tuples for pathways when we're only caring about _simple paths_ because you won't get ever get cycles. However, when you get cycles, you want tuples as opposed to sets or lists

reference for variable names
pathways: all paths from any source to any sink in a fuel cycle
path: a single path from a source to a sink
paths: all paths from a single source to a single sink. If there is more than one source and/or sink, this is not all pathways through a fuel cycle

File: trailmap/acquisition_paths.py (hunk 18 above)
Comment (nuclearkatie) [reply trailmap/acquisition_paths.py:40]: You can't += a set, nor a tuple (I just checked)

Comment (nuclearkatie) [reply trailmap/acquisition_paths.py:40]: There are a few comment threads about the same lines of code, so I'm going to tag you into the response I just wrote to Paul, and I wrote a similar response further up. 

tl;dr the output of `nx.all_simple_paths(G, source=source, target=sink)` is a list of lists, so turning that into a tuple of lists is not the desired outcome (but I totally get where you're coming from!)

File: trailmap/acquisition_paths.py (hunk 21 above)
Comment (nuclearkatie) [reply trailmap/acquisition_paths.py:38]: @bam241 trying to consolidate all the conversation on this block of code on this comment thread

File: trailmap/acquisition_paths.py (hunk 17 above)
Comment (nuclearkatie) [reply trailmap/acquisition_paths.py:14]: I have separated the two, I think? `find_simple_paths` calculates the pathways and returns them here (line 13). `conduct_apa` (this function) just calls the methods that build the graph and find pathways and then prints the output. Are you recommending I create an additional function beyond `conduct_apa` that is used just for printing?

File: trailmap/acquisition_paths.py (hunk 18 above)
Comment (gonuke) [reply trailmap/acquisition_paths.py:40]:
```
pathways.add({ tuple(path) for path in nx.all_simple_paths(...)})
```

File: trailmap/acquisition_paths.py (hunk 21 above)
Comment (bam241) [reply trailmap/acquisition_paths.py:38]: might have some solution there: https://stackoverflow.com/questions/3523048/add-another-tuple-to-a-tuple-of-tuples

File: scripts/main.py (hunk 15 above)
Comment (nuclearkatie) [reply scripts/main.py:31]: turns out that as a positional argument (because I used `infile` as opposed to `--infile`), it's already required! So I don't need and cannot even use required=True. So I'm dropping the `if ns.infile[0] is not None:` check but leaving the parser as-is

**Comment by nuclearkatie** [comment]
Things included in the last commit
- moved global imports first
- removed any code referencing or using `pathway_analysis`, which was removed from this PR
- recognized `infile` as a positional argument in argparse which automatically means it is required, and thus does not need an extra check in `main` confirming it exists
- moved pathways printing to its own function, and also added a command line flag to suppress printing
- simplified the edge adding process in `build_graph` using @gonuke's recommendation
- reduced the process in finding simple paths taking advantage of discussions with @gonuke and @bam241

**Review by gonuke — APPROVED** [review]

**Comment by gonuke** [comment]
Thanks @nuclearkatie  - LGTM!
